### PR TITLE
Migrate to azure-blob

### LINKED ISSUE #30352: Tracking issue RE: Paperclip, Azure EOL, rest client, etc
### Background

- For file attachment on various DB models throughout the app, Mastodon relies on paperclip
- With the introduction of Rails activestorage solution, the original paperclip gem was [deprecated in 2018](https://thoughtbot.com/blog/closing-the-trombone)
- There is a [kt-paperclip](https://rubygems.org/gems/kt-paperclip) fork that is more or less maintained, and which Mastodon now uses instead of paperclip
- We support various storage backends, including Azure
- For azure specifically, we rely on [md-paperclip-azure](https://github.com/mistydemeo/paperclip-azure) which is itself a fork of the orignal paperclip-azure gem (not updated in ~5 years) and provides an interface between paperclip and the official azure gem.

### Issues

- The situation where we are using a fork of an unmaintained adapter (azure) to interface with a fork of an unmaintained gem (paperclip) feels a bit precarious, but the migration from paperclip to AS would be one of the more substantial code and data/storage migrations we've attempted both on dev side and on instance admin side.
- The [azure-storage-ruby](https://github.com/Azure/azure-storage-ruby) gem will be retired in Sep 2024, in favor of using the REST api directly.
- The azure-storage-common gem is our last remaining gem with a hard version lock to faraday 1.x, which prevents updating there.

### PR DESCRIPTION
**Pull request by JoeDupuis**
`azure-storage-blob` will be [retired starting this week](https://azure.microsoft.com/en-us/updates/retirement-notice-the-azure-storage-ruby-client-libraries-will-be-retired-on-13-september-2024/)

The library (`azure-storage-blob`) hasn't been upgraded since 2021 and has a lot of dependencies. [It's already causing friction when trying to upgrade apps and dependencies.](https://github.com/rails/rails/issues/49983). 

Migrating to `azure-blob` would unblock those upgrades and close [this Mastodon issue](https://github.com/mastodon/mastodon/issues/30352).

The migration also make it possible to support managed identities instead of shared keys. I added some instructions to the readme.

This is a big change, If you prefer keeping `azure-storage-blob` as the backend, 
I can extract those changes to the azure-blob gem instead (write a new adapter).

### CODE REVIEW

File: lib/paperclip/storage/azure.rb [hunk 1]
```diff
@@ -59,13 +55,6 @@ module Storage
 
     module Azure
       def self.extended base
-        begin
-          require 'azure/storage/blob/blob_service'
-        rescue LoadError => e
-          e.message << " (You may need to install the azure-storage-blob gem)"
-          raise e
-        end unless defined?(::Azure::Core)
```
Comment (JoeDupuis) [inline lib/paperclip/storage/azure.rb:67]: This is not needed. Since the gem is in the Gemspec it will always be loaded.

**Comment by JoeDupuis** [comment]
I kept [this](https://github.com/mistydemeo/paperclip-azure/blob/master/lib/paperclip/storage/azure/environment.rb#L1-L19) because it's part of the public api, but it could probably be deleted. It's unused after the change.

File: README.md [hunk 2]
```diff
@@ -83,6 +82,18 @@ Or, in the instance where the credentials are specified at the model level:
                         region:               :cn
                       }
 
+=== Managed identities (Entra ID)
+
+To authenticate through managed identities instead of using a shared access key, omit passing the access key and pass the principal id of the identity instead. For example:
```
Comment (mistydemeo) [inline README.md:87]: Thank you for adding the docs!

File: lib/paperclip/storage/azure.rb [hunk 3]
```diff
@@ -245,32 +220,18 @@ def flush_writes #:nodoc:
...
-            block_id = "block_#{(count += 1).to_s.rjust(5, '0')}"
-
-            azure_interface.put_blob_block container_name, storage_path, block_id, data
-
-            blocks << [block_id]
-          end
-
-          azure_interface.commit_blob_blocks container_name, storage_path, blocks
-        end
+      def save_blob(storage_path, file, write_options)
```
Comment (mistydemeo) [inline lib/paperclip/storage/azure.rb:223]: Hm, guess this wasn't really part of the public interface despite not having the `:nodoc:` tag.

File: lib/paperclip/storage/azure.rb [hunk 4]
```diff
@@ -166,30 +152,26 @@ def azure_interface
       end
 
       def obtain_azure_instance_for(options)
-        if options[:use_development_storage]
```
Comment (mistydemeo) [inline lib/paperclip/storage/azure.rb:169]: Hm, I was worried about this at first, but I guess this wasn't ever documented. Dropping it with a notein the changelog seems fine.

File: lib/paperclip/storage/azure.rb [hunk 5]
```diff
@@ -166,30 +152,26 @@ def azure_interface
...
-          service = ::Azure::Storage::Blob::BlobService.create(use_development_storage: true)
-        else
-          service = ::Azure::Storage::Blob::BlobService.create(storage_account_name: options[:storage_account_name],
-                                                               storage_access_key: options[:storage_access_key])
-        end
-        
-        service
+        AzureBlob::Client.new(
+          account_name: options[:storage_account_name],
```
Comment (mistydemeo) [inline lib/paperclip/storage/azure.rb:156]: Is it fine to pass these as `nil`? I assume so.

**Review by mistydemeo — APPROVED** [review]
Thanks so much! This looks great, certainly simplifies things. I haven't had a chance to test yet, however. Have you had a chance to test in any configurations yet?

File: lib/paperclip/storage/azure.rb (hunk 3 above)
Comment (JoeDupuis) [reply lib/paperclip/storage/azure.rb:223]: Yeah :face_with_diagonal_mouth:  this one is a breaking change. It should not matter for mastodon, but I think it's worth changing major version when cutting the new release.

File: lib/paperclip/storage/azure.rb (hunk 4 above)
Comment (JoeDupuis) [reply lib/paperclip/storage/azure.rb:169]: I didn't know this was a thing before removing it here :grimacing: 
It got me thinking that I could add this functionality back in the azure-blob. Changing major version on the release should help set expectations that the API changed a little bit.

File: lib/paperclip/storage/azure.rb (hunk 5 above)
Comment (JoeDupuis) [reply lib/paperclip/storage/azure.rb:156]: `account_name` and `container` are required.
`access_key` is required if `principal_id` or `use_managed_identities` are both nil/false.
`cloud_regions` can be nil

**Comment by JoeDupuis** [comment]
> Thanks so much! This looks great, certainly simplifies things. I haven't had a chance to test yet, however. Have you had a chance to test in any configurations yet?

@mistydemeo I didn't test it with mastodon, but I did test it with a Rails app. 
I was able to upload, download, delete, use expiring url. 

Do you think it's ready to merge?

**Comment by mistydemeo** [comment]
Hm, it might be! I'd assume it'll probably work with Mastodon if it also works with a general Rails app.

I don't currently have a live production server that uses this, so I'm not in a position to test it; since you are, and you maintain the azure library that this uses, I wonder if it makes sense for you to fork my fork of paperclip-azure and publish that?

**Comment by JoeDupuis** [comment]
Just to make sure we're on the same page. 
You mean forking to test a temp release or to keep this gem as is (use a new gem instead)?

If the latter, I can extract the adapter and add it to the `azure-blob` gem directly. 
I just don't know if I'll have the time before The next Mastodon release.

**Comment by mistydemeo** [comment]
In the short term, you could fork this gem and release your own version; I don't think it's a problem if extracting the adapter and adding it to `azure-blob` is a longer-term project, since that would give an immediate migration path to anyone using the gem. (In the same way that my fork is a migration path for users of the original paperclip-azure.)

**Comment by JoeDupuis** [comment]
Released as `jd-paperclip-azure` for the Mastodon release: https://github.com/JoeDupuis/paperclip-azure
I don't have a prod mastodon instance, but the dev version worked well with the new version.

**Comment by mistydemeo** [comment]
Excellent! Should I open a PR to Mastodon to point it at the new gem, or would you like to?

Edit: I see you did already. Thank you!

**Pull request closed without merging**